finesse/hardware/st10_controller.py

```diff
@@ -0,0 +1,343 @@
+              commands. (Why the manual uses the term "exception", who knows, but it
+              also indicates success!)
+
+        If an error occurs, the device returns a negative acknowledge (NACK) response,
+        which is a "?", followed by an error code, e.g. "?4". The meanings of the error
+        codes are listed in the manual linked in the module description.
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: Error or malformed message
+        """
+        response = self._read()
+
+        # If we receive either type of ACK response
+        if response in ("%", "*"):
+            return
+
+        # An error occurred (NACK)
+        if response[0] == "?":
+            raise ST10ControllerError(
+                f"Device returned an error (code: {response[1:]})"
+            )
+
+        raise ST10ControllerError(f"Unexpected response from device: {response}")
+
+    def _request_value(self, name: str) -> str:
+        """Request a named value from the device.
+
+        You can request the values of various variables, which all seem to have
+        two-letter names.
+
+        Args:
+            name: Variable name
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: Malformed message received from device
+            UnicodeEncodeError: Message to be sent is malformed
+        """
+        self._write(name)
+        response = self._read()
+        if not response.startswith(f"{name}="):
+            raise ST10ControllerError(f"Unexpected response when querying value {name}")
+
+        return response[len(name) + 1 :]
+
+    def stop_moving(self) -> None:
+        """Immediately stop moving the motor."""
+        self._write_check("ST")
+
+    def wait_until_stopped(self, timeout: Optional[float] = None) -> None:
+        """Wait until the motor has stopped moving.
+
+        Args:
+            timeout: Time to wait for motor to finish moving (None == forever)
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for motor to finish moving
+            ST10ControllerError: Malformed message received from device
+        """
+        # Tell device to send "X" when current operations are complete
+        self._send_string("X")
+
+        # Set temporary timeout
+        old_timeout, self.serial.timeout = self.serial.timeout, timeout
+        try:
+            if self._read() != "X":
+                raise ST10ControllerError(
+                    "Invalid response received when waiting for X"
+                )
+        finally:
+            # Restore previous timeout setting
+            self.serial.timeout = old_timeout
```

Comment:
Is it the `finally` block executed even in the case of raising an exception? If not, we should restore the timeout before raising the exception.

Reply:
Yup!

In practice, if an error occurs during the read it's probably game over anyway, but it seems prudent to reset the timeout.